jupyterlab/handlers/yjs_echo_ws.py

```diff
@@ -25,170 +25,163 @@
 
 YFILE = YDOCS["file"]
 
-ROOMS = {}
+RENAME_SESSION = 127
 
-# The y-protocol defines messages types that just need to be propagated to all other peers.
-# Here, we define some additional messageTypes that the server can interpret.
-# Messages that the server can't interpret should be broadcasted to all other clients.
 
+class JupyterRoom(YRoom):
+    def __init__(self, type):
+        super().__init__(has_internal_ydoc=True)
+        self.type = type
+        self.cleaner = None
+        self.watcher = None
+        self.document = YDOCS.get(type, YFILE)(self.ydoc)
```

Comment:
Can you please walk me through this python sorcery? How does it initialize a Yjs document exactly?

Reply:
[Jupyter Y documents](https://github.com/davidbrochart/jupyter_ydoc) are registered through [entry points](https://github.com/davidbrochart/jupyter_ydoc/blob/3ec7528c8c5d714bbcdabcd7f9ec4305f891c4d4/setup.cfg#L30-L33). We currently have two built-in ydocs:
- [YFile](https://github.com/davidbrochart/jupyter_ydoc/blob/3ec7528c8c5d714bbcdabcd7f9ec4305f891c4d4/jupyter_ydoc/ydoc.py#L49-L70): a generic text document.
- [YNotebook](https://github.com/davidbrochart/jupyter_ydoc/blob/3ec7528c8c5d714bbcdabcd7f9ec4305f891c4d4/jupyter_ydoc/ydoc.py#L73-L155): a Jupyter notebook.

They are high-level documents that use Ypy's `YDoc` underneath, but with methods to get text/JSON source, set from text/JSON, and observe changes. They must of course mimic their equivalent structures in the front-end.
Here `YDOCS` is just a dictionary of all the available Jupyter documents, e.g.:
```py
{'file': <class 'jupyter_ydoc.ydoc.YFile'>, 'notebook': <class 'jupyter_ydoc.ydoc.YNotebook'>}
```
We create a Jupyter document corresponding to the type requested by the front-end, defaulting to a `YFile`.

Reply:
Thank you, this is exactly what I was looking for!

It makes a lot of sense to outsource this to a separate package. This is basically the equivalent of `shared-models` package in JupyterLab. What do you think about naming this package `shared-models` as well?

Reply:
I'm not sure, `shared-model` is a good name in the context of JupyterLab, but the current `jupyter_ydoc` could be used outside JupyterLab. For instance, I'm planning to use it in [jpterm](https://github.com/davidbrochart/jpterm), a Python Jupyter client that allows to work collaboratively with e.g. JupyterLab:

https://user-images.githubusercontent.com/4711805/166236284-320b5275-79c0-4a76-9aa5-e2ad09964812.mp4

: